Ea 2577 solr indexing

SrdjanStevanetic opened the pull request:
Please review the code.

SrdjanStevanetic committed:
added a maven module for solr indexing

SrdjanStevanetic committed:
solr indexing test and merge branch 'develop' into EA-2577-solr-indexing

Conflicts:
	entity-management-common/src/main/java/eu/europeana/entitymanagement/common/config/EntityManagementConfiguration.java
	pom.xml

SrdjanStevanetic committed:
the suggester field is added to the solr entities

SrdjanStevanetic committed:
Merge branch 'develop' into EA-2577-solr-indexing

Conflicts:
	entity-management-web/src/main/java/eu/europeana/entitymanagement/web/EMController.java

SrdjanStevanetic committed:
removed org.json from the solr .pom file

SrdjanStevanetic committed:
commented out the solr test

ikattey commented on entity-management-definitions/src/main/java/eu/europeana/entitymanagement/definitions/exceptions/EntityCreationException.java:
```diff
@@ -32,8 +32,8 @@ public boolean logStacktrace() {
+//    @Override
+//    public HttpStatus getResponseStatus() {
+//        return HttpStatus.BAD_REQUEST;
+//    }
```
Error responses with this exception currently return a 400 status code.

If this method is removed, a 500 response will be returned instead. See `createEntityFromMigrationRequest()` and `createEntityFromRequest()` in `EntityRecordService`

It's fine to have this return a 500; however it needs to be replaced by a more suitable exception in `EntityRecordService`.

SrdjanStevanetic replied:
I reverted back the 400 status code.

ikattey commented on entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/exception/SolrServiceException.java:
```diff
@@ -0,0 +1,16 @@
+package eu.europeana.entitymanagement.solr.exception;
+
+public class SolrServiceException extends Exception {
```
By convention, all custom exceptions should extend `EuropeanaApiException`

SrdjanStevanetic replied:
Improved.

ikattey commented on entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/model/SolrAgent.java:
```diff
@@ -0,0 +1,241 @@
+/*
+ * TODO:see how to save the referencedWebResource and isAggregatedBy fields for all entities
+ */
+public class SolrAgent extends Agent {
```
Consider making this a different type. See the implementation in `XmlAgentImpl`. 

This comment applies to all `Solr*` classes.

SrdjanStevanetic replied:
I think this is not necessary. We need to have the same fields in the Solr entities as in the main entity classes, and simply extending from the main classes keeps that validation in place. We also do not need to define the new fields in the Solr classes, simply use the @Field annotation on the setters for the existing fields. However, we need to define the base entity fields in all Solr classes independently and cannot keep them in one class, but it is fine I guess.

ikattey replied:
There's a lot of duplicated code in the Solr model classes, which can be eliminated with a different class hierarchy.

See [here](https://github.com/europeana/entity-management/compare/EA-2577-solr-indexing...EA-2577-solr-entities#diff-a0bb5be85e98ff1bec4215d9d8ba320a3b044302cccce6a5f1e6eda08a5250e8R11) for a demonstration of what I'm recommending.

SrdjanStevanetic replied:
Improved.

ikattey commented on entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/service/SolrService.java:
```diff
@@ -0,0 +1,19 @@
+	 * @param doCommit commit
+	 * @throws SolrServiceException 
+	 */
+	public void storeEntity(Entity solrObject, boolean doCommit) throws SolrServiceException ;
```
Method signature makes it very easy to store the wrong objects in Solr.

Consider introducing a new type for Solr, similar to what we do for XML handling.

SrdjanStevanetic replied:
Improved.

ikattey commented on entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/service/SolrService.java:
```diff
@@ -0,0 +1,19 @@
+import eu.europeana.entitymanagement.definitions.model.Entity;
+import eu.europeana.entitymanagement.solr.exception.SolrServiceException;
+
+public interface SolrService {
```
The convention for this project is to avoid single-implementation interfaces (as mentioned on Slack).

Directly implement functionality in `SolrService`, instead of having `SolrService` interface and a `SolrServiceImpl` class.

SrdjanStevanetic replied:
Improved.

ikattey commented on entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/service/impl/SolrServiceImpl.java:
```diff
@@ -0,0 +1,263 @@
+import eu.europeana.entitymanagement.vocabulary.EntityTypes;
+
+@Service(AppConfigConstants.BEAN_EM_SOLR_SERVICE)
+public class SolrServiceImpl implements SolrService {
```
As above, please implement functionality directly in `SolrService`.

SrdjanStevanetic replied:
Done.

ikattey replied:
Good! I see the class is still in the `solr/service/impl` package however 🙂

SrdjanStevanetic replied:
Improved.

ikattey commented on entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/SolrConfig.java:
```diff
@@ -0,0 +1,39 @@
+	@Bean
+	public SolrClient solrClient() {
+		logger.info("Configuring the solr client at the url: {}", emConfiguration.getSearchApiSolrUrl());
+		return new HttpSolrClient.Builder(emConfiguration.getSearchApiSolrUrl()).build();
```
Solr instance for indexing entities is different from the Search API instance.

This should be a new instance / config property.

SrdjanStevanetic replied:
Improved.

ikattey commented on entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/SolrConfig.java:
```diff
@@ -0,0 +1,39 @@
+
+@Configuration
+@EnableSolrRepositories
+public class SolrConfig {
```
All SolrClient instances (including the two in `ScoringService`) should be initialised here as named beans.

SrdjanStevanetic replied:
I have tried that and I think that Spring has some problems with using many beans with the same type even though I tried to distinguish them by using the bean names and the @Qualifier annotation while imorting them over @Autowired. Therefore I actually created the indexing SolrClient the same way as the other 2 solr clients in the ScoringService class.

ikattey replied:
See [this commit](https://github.com/europeana/entity-management/compare/EA-2577-solr-indexing...EA-2577-solr-config).

SrdjanStevanetic replied:
I see, the problem was in the SolrTemplate beans that did not work together with the SolrClient beans and you put the @Qualifier in the constructor. Fine.

ikattey reviewed (changes requested):
@SrdjanStevanetic please see my comments below.

I did not check the field names, as @gsergiu is better placed to review those.

SrdjanStevanetic committed:
improved solr indexing

ikattey committed:
Introduce generics to Solr classes

ikattey committed:
EA-2577: Configure SolrClients in config class

SrdjanStevanetic committed:
Merge branch 'EA-2577-solr-entities' into EA-2577-solr-indexing

gsergiu commented on entity-management-common/src/main/java/eu/europeana/entitymanagement/common/config/EntityManagementConfiguration.java:
```diff
@@ -28,14 +28,23 @@
     private String searchApiSolrUrl;
+    
+    @Value("${entitymanagement.solr.searchapi.collection}")
+    private String searchApiSolrCollection;
```
the collection of the search API is already part of the URL, this property is redundant

SrdjanStevanetic replied:
Removed.

gsergiu commented on entity-management-common/src/main/java/eu/europeana/entitymanagement/common/config/EntityManagementConfiguration.java:
```diff
@@ -28,14 +28,23 @@
+    private String indexingSolrUrl;
+    
+    @Value("${entitymanagement.solr.indexing.collection}")
+    private String indexingSolrCollection;
```
this property is redundant, the collection should be part of the solr URL

SrdjanStevanetic replied:
Removed.

gsergiu commented on entity-management-solr/pom.xml:
```diff
@@ -0,0 +1,65 @@
+  <name>entity-management-solr</name>
+  <description>A simple entity-management-solr.</description>
+  <!-- FIXME change it to the project's website -->
+  <url>http://www.example.com</url>
```
please remove this line

SrdjanStevanetic replied:
Removed.

gsergiu commented on entity-management-solr/pom.xml:
```diff
@@ -0,0 +1,65 @@
+  <!-- FIXME change it to the project's website -->
+  <url>http://www.example.com</url>
+
+  <properties>
```
these properties are inherited from the base pom, please remove

SrdjanStevanetic replied:
Removed.

gsergiu commented on entity-management-solr/pom.xml:
```diff
@@ -0,0 +1,65 @@
+   
+  </build>
+
+  <reporting>
```
do we need this plugin? If not, please remove

SrdjanStevanetic replied:
Removed.

gsergiu commented on entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/model/SolrAgent.java:
```diff
@@ -0,0 +1,99 @@
+		super(agent);
+
+		
+		this.date = agent.getDate();
```
aren't these values already set in the super(agent) ? Only the multilingual maps need preProcessing to append solrFieldName to labels, but that should also be applied automatically through the call of the supper implementation

SrdjanStevanetic replied:
The new solr classes are implemented.

gsergiu commented on entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/model/SolrConcept.java:
```diff
@@ -0,0 +1,199 @@
+	}
+
+	public SolrConcept(Concept concept) {
+		super();
```
please switch to the use of super(concept) and remove the redundant calls to set the values.

SrdjanStevanetic replied:
Improved.

gsergiu commented on entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/model/SolrEntity.java:
```diff
@@ -0,0 +1,142 @@
+
+public abstract class SolrEntity<T extends Entity> {
+
+    @Field(AgentSolrFields.TYPE)
```
This is a common field, please use the EntitySolrFields class instead

SrdjanStevanetic replied:
Done.

gsergiu commented on entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/model/SolrEntity.java:
```diff
@@ -0,0 +1,142 @@
+    @Field(AgentSolrFields.TYPE)
+    private String type;
+
+    @Field(AgentSolrFields.ID)
```
This is a common field, please use the EntitySolrFields class insted

SrdjanStevanetic replied:
Done.

gsergiu commented on entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/model/SolrEntity.java:
```diff
@@ -0,0 +1,142 @@
+package eu.europeana.entitymanagement.solr.model;
+
+import eu.europeana.entitymanagement.definitions.model.Entity;
+import eu.europeana.entitymanagement.vocabulary.AgentSolrFields;
```
Please remove this import and use the constants from EntitySolrFields class instead

SrdjanStevanetic replied:
Done.

gsergiu commented on entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/model/SolrOrganization.java:
```diff
@@ -0,0 +1,265 @@
+	}
+
+	public SolrOrganization(Organization organization) {
+		super();
```
use super(organization) and remove the use of redundant setters

SrdjanStevanetic replied:
Done.

gsergiu commented on entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/model/SolrPlace.java:
```diff
@@ -0,0 +1,168 @@
+	}
+
+	public SolrPlace(Place place) {
+		super();
```
use super(place) and remove the use of redundant setters

SrdjanStevanetic replied:
Done.

gsergiu commented on entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/model/SolrTimespan.java:
```diff
@@ -0,0 +1,146 @@
+	}
+
+	public SolrTimespan(Timespan timespan) {
+		super();
```
use super(timespan) and remove the use of redundant setters

SrdjanStevanetic replied:
Done.

gsergiu commented on entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/service/SolrService.java:
```diff
@@ -0,0 +1,186 @@
+
+	@Qualifier(AppConfigConstants.BEAN_JSON_MAPPER)
+	@Autowired
+	ObjectMapper emJsonMapper;	
```
if this is the mapper used for serialization of the preview, please rename to payloadObjectMapper

SrdjanStevanetic replied:
Done.

gsergiu commented on entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/utils/SolrEntityUtils.java:
```diff
@@ -0,0 +1,20 @@
+import eu.europeana.entitymanagement.solr.model.SolrAgent;
+import eu.europeana.entitymanagement.solr.model.SolrEntity;
+
+public class SolrEntityUtils {
```
this class doesn't seem to be needed, if not used please remove

SrdjanStevanetic replied:
Removed.

gsergiu commented on entity-management-solr/src/site/site.xml:
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
for what is this file needed?

SrdjanStevanetic replied:
Removed.

gsergiu commented on entity-management-web/src/integration-test/java/eu/europeana/entitymanagement/solr/SolrServiceTest.java:
```diff
@@ -0,0 +1,68 @@
+    JsonLdSerializer emJsonldSerializer;
+
+    //@Test
+    public void storeEntitiesInSolr() throws Exception {
```
please create one test for each entity class

SrdjanStevanetic replied:
Done.

gsergiu commented on entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/service/SolrService.java:
```diff
@@ -0,0 +1,186 @@
+	public void storeEntity(final SolrEntity<? extends Entity> solrEntity, boolean doCommit) throws SolrServiceException {
+
+		// TODO: adapt and uncomment
+//			try {
```
this is that EA-2578 has to implement.

SrdjanStevanetic replied:
Ok.

gsergiu reviewed (changes requested):
please implement the required changes and complete serialization of payload field

SrdjanStevanetic committed:
Merge branch 'EA-2577-solr-config' into EA-2577-solr-indexing

Conflicts:
	entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/service/impl/SolrService.java

SrdjanStevanetic committed:
Json filter for the Solr suggester field added and the corresponding
code adaptations

SrdjanStevanetic committed:
setting up the FilterProvider for the JsonFilter annotation in the main
json serializer

SrdjanStevanetic committed:
some improvements in the code for the Solr module

SrdjanStevanetic committed:
commented out the solr tests

ikattey committed:
EA-2577: Setup Solr Testcontainer for integration tests (#82)

ikattey committed:
EA-2577: Improve Solr indexing

- introduced config property for enabling explicit commits
- created Bean for SolrEntity FilterProvider

ikattey committed:
EA-2577: Index entities to Solr during registration and update task

ikattey committed:
EA-2577: Fix SolrService integration test

ikattey committed:
EA-2577: Correctly generate Solr suggester payload

ikattey committed:
Merge branch 'develop' into EA-2577-solr-indexing

Commit:
#EA-2577 #EA-2622 updated solr configurations to be in sync with the
ones for test server

ikattey committed:
Merge branch 'develop' into EA-2577-solr-indexing

Commit:
#EA-2622 enable softautocommit, removed unused property

Commit:
Merge remote-tracking branch 'origin/develop' into EA-2577-solr-indexing

Conflicts:
	entity-management-definitions/src/main/java/eu/europeana/entitymanagement/definitions/model/Entity.java
